[doyelee0313] iP

doyelee0313 opened the pull request.

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

doyelee0313 committed:
Level-0

doyelee0313 committed:
Level-1

doyelee0313 committed:
Level-2

doyelee0313 committed:
Level-3

doyelee0313 committed:
Level-4

doyelee0313 committed:
A-TextUiTesting

doyelee0313 committed:
A-Exceptions

doyelee0313 committed:
Level-6

doyelee0313 committed:
javadocs

doyelee0313 committed:
add-intro

doyelee0313 committed:
Revert "add-intro"

This reverts commit 47305a7558a015e507de8cf369d5398f945d6009.

doyelee0313 committed:
branch-Level-7

doyelee0313 committed:
merge commit

doyelee0313 committed:
Merge branch 'master' into branch-Level-7

doyelee0313 committed:
Level-8

doyelee0313 committed:
oop start

doyelee0313 committed:
Revert "oop start"

This reverts commit 9ba68a7a3742d5e86d52658087b8b0f9139026b5.

doyelee0313 committed:
oop start

doyelee0313 committed:
A-MoreOOP

doyelee0313 committed:
A-Packages

doyelee0313 committed:
Merge branch 'branch-A-MoreOOP'

doyelee0313 committed:
Merge branch 'master' into add-gradle-support

doyelee0313 committed:
A-Gradle

doyelee0313 committed:
Merge branch 'add-gradle-support'

doyelee0313 committed:
A-JUnit

doyelee0313 committed:
A-JUnit

doyelee0313 committed:
A-JavaDoc

doyelee0313 committed:
A-CodingStandard

doyelee0313 committed:
Level-9

doyelee0313 committed:
Merge tag 'A-CodingStandard'

doyelee0313 committed:
Merge tag 'Level-9'

# Conflicts:
#	src/main/java/duke/Parser.java

doyelee0313 committed:
Update README.md

doyelee0313 committed:
Level-10

doyelee0313 committed:
Level-10

doyelee0313 committed:
Merge branch 'master' of https://github.com/doyelee0313/ip

Kaya3842 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,108 @@
+ */
+public class TaskList {
+    
+    private List<Task> tasks;
+
+    /**
+     * Constructs a TaskList instance.
+     *
+     * @param tasks The list of tasks to initialize the TaskList.
+     */
+    public TaskList(List<Task> tasks) {
+        this.tasks = tasks;
+    }
+
+    /**
+     * Constructs an empty TaskList.
+     */
+    public TaskList() {
+        this.tasks = new ArrayList<>();
+    }
```
Why not use only ArrayList<Task>?

Kaya3842 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,108 @@
+        return tasks.get(number - 1);
+    }
+
+    /**
+     * Unmarks a task.
+     *
+     * @param number The number of the task to be unmarked.
+     */
+    public void unmark(int number) {
+        Task t = tasks.get(number - 1);
+        t.markAsUnDone();
+    }
+
+    /**
+     * Marks a task as done.
+     *
+     * @param number The number of the task to be marked as done.
+     */
+    public void mark(int number) {
+        Task t = tasks.get(number - 1);
+        t.markAsDone();
+    } 
```
Shouldn't these two methods belong to the Task class?

Kaya3842 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,145 @@
+    private Ui ui;
+    private Parser parser;
+
+    private static String filePath = "/Users/leedoye/ip/src/data/duke_tasks.txt";
```
I think the file path should not be absolute.

Kaya3842 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,54 @@
+ *
+ */
+public class Storage {
+    private String filePath = "/Users/leedoye/ip/src/data/duke_tasks.txt";
```
Same issue regarding absolute filepath.

Kaya3842 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,54 @@
+     * @return The list of tasks loaded from the file.
+     * @throws DukeException If there is an error loading tasks from storage.
+     */
+    public List<Task> load() throws DukeException{
+        try (ObjectInputStream intput = new ObjectInputStream(new FileInputStream("filePath"))) {
+            return (List<Task>) intput.readObject();
+        } catch (IOException | ClassNotFoundException e) {
+            System.out.println("No existing tasks or file.");
+            throw new DukeException("Error loading tasks from storage");
+        }
+    }
```
```suggestion
    public List<Task> load() throws DukeException{
        try {
            ObjectInputStream intput = new ObjectInputStream(new FileInputStream("filePath"));
            return (List<Task>) intput.readObject();
        } catch (IOException | ClassNotFoundException e) {
            System.out.println("No existing tasks or file.");
            throw new DukeException("Error loading tasks from storage");
        }
    }
```

Kaya3842 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,54 @@
+     * @return The list of tasks saved from the file.
+     * @throws DukeException If there is an error saving tasks from storage.
+     */
+    public void save(List<Task> tasks) {
+        try (ObjectOutputStream output = new ObjectOutputStream(new FileOutputStream("filePath"))) {
+            output.writeObject(tasks);
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
```
Same issue with try-catch statements can be found more in this class.

Kaya3842 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,145 @@
+            try {
+                Task task = parser.parseDeadline(input);
+                tasks.addTask(task);
+                response = ui.addedMessage(task) + ui.totalTask(tasks.getSize());
+                storage.save(tasks.getTasks());
+            } catch (DukeException e) {
+                response = e.getMessage();
+            }
+        } else if (orders[0].equals("event")) { 
+            try {
+                Task task = parser.parseEvent(input);
+                tasks.addTask(task);
+                response = ui.addedMessage(task) + ui.totalTask(tasks.getSize());
+                storage.save(tasks.getTasks());
+            } catch (DukeException e) {
+                response = e.getMessage();
+            }
+        } else if (orders[0].equals("delete")) {
+            try {
+                int number = parser.parseDelete(input, tasks.getSize());
+                response = ui.deletedMessage(tasks.getTask(number));
+                tasks.deleteTask(number);
+                response += ui.totalTask(tasks.getSize());
+                storage.save(tasks.getTasks());
+            } catch (DukeException e) {
+                response = e.getMessage();
+            }
+        } else if (orders[0].equals("find")) {
+            try {
+                String word = parser.parseFind(input);
+                TaskList newtasks = new TaskList(tasks.findTasks(word));
+                response = ui.findList(newtasks.getTasks());
+            } catch (DukeException e) {
+                response = e.getMessage();
+            }
+        } else {
+            response = ui.dontUnderstand();
+        }
+        return response;
+    }
```
Should this be extracted out? For example you can extract one method for Parser class and one method for Ui class instead of combining them in the Duke class. Or create a new method only for extracting the numbers in a command.

Kaya3842 reviewed (commented):
Most of the classes was extracted out reasonably well. There was some minor coding standard violations but it can easily be fixed.  Consider extracting more methods in Parser and Ui class to make the main Duke class shorter.

doyelee0313 committed:
Enhance error handling in the Parser class by adding assert statements to verify the validity of input parameters.

It needs to change to ensure that critical assumptions about input parameters are validated during runtime, helping to catch potential issues and improve code reliability.

Add assert statements to validate input parameters in various parsing methods.

This approach enhances code robustness by explicitly documenting and enforcing important assumptions about input parameters.

The use of assert statements helps catch errors early during development and testing.

Assertions are are primarily intended for use during development and testing.

doyelee0313 committed:
Change: Updated comments in Task class for clarity and consistency.

Improvement: The comments were updated to provide clearer explanations.

Action: Update comments in Task class.

Reason: Clear and consistent comments improve code readability and maintainability.

Additional Info: No functional changes were made, only comments were updated.

doyelee0313 committed:
Merge branch 'A-CodeQuality'

doyelee0313 committed:
BCD-Extension

doyelee0313 committed:
Update README.md

doyelee0313 committed:
Update README.md

doyelee0313 committed:
Update README.md

doyelee0313 committed:
A-UserGuide

doyelee0313 committed:
Merge commit '962f1f916886982419c12f2ef8cdb18511a44805'

doyelee0313 committed:
A-Release

doyelee0313 committed:
final version

doyelee0313 committed:
Update README.md

doyelee0313 committed:
Update README.md

doyelee0313 committed:
Update README.md

doyelee0313 committed:
Update README.md

doyelee0313 committed:
Update README.md

doyelee0313 committed:
Update README.md
